### ISSUE TITLE
Add closure variant of `SignalBuilder::class_handler()` and use `Value` traits for `accumulator()`

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gtk-rs-core#365]

**Comment by sdroege** [comment]
`accumulator()` should probably also be based on `std::ops::ControlFlow` at this point.

**Comment by sdroege** [comment]
This one is not so easy because of the `SignalClassHandlerToken` we have to safely allow chaining up

**Comment by sdroege** [comment]
Class handler is: `Fn(&SignalClassHandlerToken, &[Value]) -> Option<Value> + Send + Sync + 'static`

If we make this a `GClosure` then somehow the token has to be passed in there.

@jf2048 Do you maybe have an idea?

Or @bilelmoussaoui @GuillaumeGomez ? :)

----

Accumulator is: `Fn(&SignalInvocationHint, &mut Value, &Value) -> bool + Send + Sync + 'static`

This has no `GClosure` API but that's fine, we can directly make this a `T: FromValue + ToValue, Fn(&SignalInvocationHint, T) -> ControlFlow<Value, T> + Send + Sync + 'static`

**Comment by jf2048** [comment]
Sorry I missed this before, I can't think of a good way to do this without creating a second `GClosure` to wrap the one passed in. If you want to avoid that maybe you have to accept that using your own `GClosure` and chaining is unsafe. The easier way to do this and still get the value unwrapping is to have a real [signals macro](https://github.com/gtk-rs/gtk-rs-core/issues/214), which I have implemented and it works the way you would expect. The accumulator changes look good to me.

BTW it seems it may be possible to remove `SignalClassHandlerToken` and replace it with a thread local that stores a stack of the current signal emissions, similar to what gobject is actually doing underneath the hood.

**Comment by jf2048** [comment]
Also the way this is implemented right now makes no sense to me, the top-level class closure (registered with `g_signal_newv`) definitely doesn't need the token, so that can be removed from `class_handler`

**Comment by sdroege** [comment]
You need that token to pass it along for chaining up as otherwise you could chain up from everywhere and that's not allowed

**Comment by jf2048** [comment]
I don't think that is any worse than being allowed to call any of the `parent_*` functions from everywhere which you can technically do right now

**Comment by sdroege** [comment]
Thread local tracking has runtime performance impact though

**Comment by jf2048** [comment]
I don't think it should have any more performance impact than what is happening now, it's essentially just taking the token and putting it on a thread local. It should still be better than the code inside gobject which always needs to lock the global mutex on all signal emissions :frowning_face:

What I mean though is you never chain up from the class handler, only from an overridden class handler. It's not needed in `Signal::class_handler`, only in `ObjecClassSubclassExt::override_signal_class_handler`. I think it is safe to just make a version of `class_handler` that takes a closure and drop the token from there, if that is good enough.

If we stick with the token there are some changes that should be probably be made, e.g.

- The token should probably be passed by value and get consumed by calling `signal_chain_from_overridden`, right now you can call that multiple times on the same token which is bad
- There should probably be some way to change the arguments when chaining up because this is desirable in some cases, but this requires type checking the `Value`s at run time to be type safe. Eventually in a signals macro I'd like to autogenerate a `chain_*` function for overridden handlers, and that won't need type checking

**Comment by sdroege** [comment]
> I don't think it should have any more performance impact than what is happening now

Yeah you're probably right. I'm not opposed to changing this in any case if it solves some actual problem :)



> I think it is safe to just make a version of `class_handler` that takes a closure and drop the token from there, if that is good enough.

I think you're right, that must've been some confusion that caused me to end up with the current API :)
